internal/tempfile/tmpfile_test.go

```diff
@@ -0,0 +1,173 @@
+package tempfile_test
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/buildkite/agent/v3/internal/tempfile"
+	"gotest.tools/v3/assert"
+)
+
+func TestNew(t *testing.T) {
+	t.Parallel()
+
+	for _, test := range []struct {
+		name  string
+		opts  []tempfile.TempFileOpts
+		check func(t *testing.T, f *os.File)
+	}{
+		{
+			name:  "default",
+			opts:  []tempfile.TempFileOpts{},
+			check: func(t *testing.T, f *os.File) {},
+		},
+		{
+			name: "with filename",
+			opts: []tempfile.TempFileOpts{
+				tempfile.WithName("foo.txt"),
+			},
+			check: func(t *testing.T, f *os.File) {},
+		},
+		{
+			name: "with dir",
+			opts: []tempfile.TempFileOpts{
+				tempfile.WithDir("foo"),
+			},
+			check: func(t *testing.T, f *os.File) {
+				assert.Check(t, strings.HasPrefix(f.Name(), filepath.Join(os.TempDir(), "foo")))
+			},
+		},
+		{
+			name: "with perms",
+			opts: []tempfile.TempFileOpts{
+				tempfile.WithPerms(0o600),
+			},
+			check: func(t *testing.T, f *os.File) {
+				if runtime.GOOS == "windows" {
+					t.Skip("Windows doesn't support or need checking if chmod worked")
+				}
+
+				fi, err := os.Stat(f.Name())
+				assert.NilError(t, err, "os.Stat(%q) = %s", f.Name(), err)
+				assert.Check(t, fi.Mode().Perm() == os.FileMode(0o600))
+			},
+		},
+		{
+			name: "with filename and keep extension",
+			opts: []tempfile.TempFileOpts{
+				tempfile.WithName("foo.txt"),
+				tempfile.KeepingExtension(),
+			},
+			check: func(t *testing.T, f *os.File) {
+				assert.Check(t, filepath.Ext(f.Name()) == ".txt")
+			},
+		},
+		{
+			name: "without filename and keep extension",
+			opts: []tempfile.TempFileOpts{
+				tempfile.KeepingExtension(),
+			},
+			check: func(t *testing.T, f *os.File) {},
+		},
+	} {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			f, err := tempfile.New(test.opts...)
+			assert.NilError(t, err, `New(%v) = %v`, test.opts, err)
+
+			t.Cleanup(func() {
+				assert.NilError(t, f.Close(), "failed to close file: %s", f.Name())
```

Comment:
`assert.NilError` uses `t.FailNow`, which stops the test immediately (https://go.dev/play/p/4QnMdl90ZUn). 

But since this is cleanup code - we should probably try to `os.Remove` the file whether or not `f.Close` succeeded.

Reply:
They are `assert.Check`s now. We could also use multiple `t.Cleanup`s, but it gets a bit confusing b/c they're LIFO. It's no more complicated than `defer` thought, but in my experience, it's more common to write multi-statement `defer`s than to write multiple consecutive `defer` statements.